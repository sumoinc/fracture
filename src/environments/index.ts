export * from "./environment";
export * from "./region";
