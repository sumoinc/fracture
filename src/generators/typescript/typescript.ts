import { SourceCode } from "projen";
import { FractureComponent } from "../../core/component";
import { Fracture } from "../../core/fracture";

export class TypeScript extends FractureComponent {
  constructor(fracture: Fracture) {
    super(fracture);
  }

  public preSynthesize() {
    /**
     * Build a TypeScript file and make sure it's marked as being managed by
     * projen.
     *
     * @param name
     * @returns {SourceCode}
     */
    const typeScriptFile = (name: string): SourceCode => {
      const f = new SourceCode(
        this.project,
        `${this.fracture.gendir}/types/${name}`
      );
      f.line("// " + f.marker);
      f.line("\n");
      return f;
    };

    // create index file
    const indexFile = typeScriptFile("index.ts");

    /**
     * Generate types for each entity.
     */
    this.fracture.entities.forEach((e) => {
      // add to index for export
      indexFile.line(`export * from "./${e.name}";`);

      // generate types for this entity
      const tsFile = typeScriptFile(`${e.name}.ts`);

      // generate helpful comments
      tsFile.line(`/**`);
      e.comment.forEach((c) => tsFile.line(` * ${c}`));
      tsFile.line(` */`);

      // build interface
      tsFile.open(`export interface ${e.name} {`);
      e.attributes.forEach((a) => {
        // generate helpful comments
        tsFile.line(`/**`);
        a.comment.forEach((c) => tsFile.line(` * ${c}`));
        tsFile.line(` */`);
        // console.log(a);

        // write type definition
        const q = a.isRequired ? "" : "?";
        tsFile.line(`${a.name}${q}: ${a.typeScriptType};`);
      });
      tsFile.close(`}`);
      tsFile.line("\n");
    });
    // close up index file
    indexFile.line("\n");
  }
}
