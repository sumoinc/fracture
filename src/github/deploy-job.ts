import { Component } from "projen";
import {
  Job,
  JobCallingReusableWorkflow,
  JobPermission,
} from "projen/lib/github/workflows-model";
import { Fracture } from "../core";

export interface DeployJobOptions {
  /**
   * Job ID for the deploy job.
   *
   * @default deploy
   */
  jobId: string;
  /**
   * The command to run to deploy the code.
   *
   * @default "npx projen deploy"
   */
  deployCommand: string;
}

export class DeployJob extends Component {
  /**
   * The job definition, suitable for adding to a workflow.
   */
  public readonly job: Job | JobCallingReusableWorkflow;
  /**
   * The Job Id for the deploy job
   *
   * @default deploy
   */
  public readonly jobId: string;
  /**
   * The command to run to deploy the code.
   *
   * @default "npx projen deploy"
   */
  public readonly deployCommand: string;

  constructor(fracture: Fracture, options: Partial<DeployJobOptions> = {}) {
    super(fracture);

    /***************************************************************************
     * Props
     **************************************************************************/

    this.jobId = options.jobId ?? "deploy";
    this.deployCommand = options.deployCommand ?? "npx projen deploy";

    /***************************************************************************
     * Create job definition
     **************************************************************************/

    this.job = {
      runsOn: ["ubuntu-latest"],
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.WRITE,
      },
      steps: [
        {
          name: "Deploy",
          run: this.deployCommand,
        },
      ],
    };
  }
}
