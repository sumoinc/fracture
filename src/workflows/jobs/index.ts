export * from "./build-job";
export * from "./deploy-job";
