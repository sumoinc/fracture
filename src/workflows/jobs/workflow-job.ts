import { Component } from "projen";
import { Job } from "projen/lib/github/workflows-model";
import { NodeProject } from "projen/lib/javascript";

export interface WorkflowJobOptions {
  /**
   * Job's Id
   */
  readonly jobId: string;
  /**
   * Job's name
   */
  readonly name: string;
}

export class WorkflowJob extends Component {
  /**
   * Build Job's Id for this workflow.
   */
  public readonly jobId: string;
  /**
   * Job's name
   */
  public readonly name: string;

  constructor(
    public readonly project: NodeProject,
    options: WorkflowJobOptions
  ) {
    super(project);

    //defaults
    this.jobId = options.jobId;
    this.name = options.name;
  }

  render = (): Job => {
    throw new Error("Not Implemented");
  };
}
