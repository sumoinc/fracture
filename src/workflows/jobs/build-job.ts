import { Job, JobPermission } from "projen/lib/github/workflows-model";
import { SetOptional } from "type-fest";
import { WorkflowJob, WorkflowJobOptions } from "./workflow-job";
import { renderBuildSteps } from "../steps/build-steps";
import { renderUploadArtifactSteps } from "../steps/upload-artifact-steps";
import { Workflow } from "../workflow";

export interface BuildJobOptions
  extends SetOptional<WorkflowJobOptions, "jobId" | "name"> {}

export class BuildJob extends WorkflowJob {
  constructor(
    public readonly workflow: Workflow,
    options: BuildJobOptions = {}
  ) {
    super(workflow, {
      jobId: "build",
      name: "Build",
      ...options,
    });
  }

  render = (): Job => {
    return {
      name: "Build",
      runsOn: this.workflow.defaultRunners,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.WRITE,
      },
      steps: [
        ...renderBuildSteps(this.workflow),
        ...renderUploadArtifactSteps(this),
      ],
    };
  };
}
