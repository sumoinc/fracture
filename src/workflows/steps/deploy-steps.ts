import { JobStep } from "projen/lib/github/workflows-model";
import { AwsEnvironment } from "../../environments";
import { DeployJob } from "../jobs/deploy-job";

export const renderDeploySteps = (deployJob: DeployJob): Array<JobStep> => {
  const steps: Array<JobStep> = [];

  // if deploying top AWS, configure credentials
  if (deployJob.environment.authProviderType === "AWS_GITHUB_OIDC") {
    const e = deployJob.environment as AwsEnvironment;
    steps.push({
      name: "Configure AWS Credentials",
      uses: "aws-actions/configure-aws-credentials@v2",
      with: {
        "role-to-assume": `arn:aws:iam::${e.accountNumber}:role/${e.gitHubDeploymentOIDCRoleName}`,
        "aws-region": e.region,
        "role-duration-seconds": e.gitHubDeploymentOIDCRoleDurationSeconds,
      },
    });
  }
  /*
  authProviderType: awsEnvironment.authProviderType,
      awsCredentialsOidc: {
        roleToAssume: ,
        roleDurationSeconds:
          awsEnvironment.gitHubDeploymentOIDCRoleDurationSeconds,
        awsRegion: awsEnvironment.region,
      },
  */

  deployJob.deploySteps.forEach((step) => {
    steps.push(step);
  });

  return steps;
};
