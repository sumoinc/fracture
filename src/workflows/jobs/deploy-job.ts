import { paramCase } from "change-case";
import { Job, JobPermission, JobStep } from "projen/lib/github/workflows-model";
import { TypeScriptProject } from "projen/lib/typescript";
import { SetOptional } from "type-fest";
import { WorkflowJob, WorkflowJobOptions } from "./workflow-job";
import { Environment } from "../../environments";
import { renderDeploySteps } from "../steps/deploy-steps";
import { renderDownloadArtifactSteps } from "../steps/download-artifact-steps";
import { renderSetupNodeSteps } from "../steps/setup-node-steps";
import { Workflow } from "../workflow";

export interface DeployJobOptions
  extends SetOptional<WorkflowJobOptions, "jobId" | "name"> {
  /**
   * How to identify the app being deployed
   *
   */
  readonly appName: string;

  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  readonly branchPrefix?: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  readonly needs?: Array<string>;

  /**
   * Step definition(s) that conduct the deployment
   */
  readonly deploySteps: Array<JobStep>;

  /**
   * Environment deploying to.
   */
  readonly environment: Environment;

  /**
   * Directory that was preserved for this project in order to deploy
   * properly in deploy step.
   */
  readonly artifactsDirectory: string;
}

export class DeployJob extends WorkflowJob {
  /**
   * The branch pattern this deployment is targeting.
   *
   * @default main
   */
  public readonly branchPrefix: string;

  /**
   * Jobs that must complete before this one may run
   *
   * @default - the build job
   */
  public readonly needs: Array<string>;

  /**
   * Step definition(s) that conduct the deployment
   */
  public readonly deploySteps: Array<JobStep>;

  /**
   * Environment deploying to.
   */
  public readonly environment: Environment;

  /**
   * Directory that was preserved for this project in order to deploy
   * properly in deploy step.
   */
  public readonly artifactsDirectory: string;

  /**
   * Workflow for deployment
   */
  readonly workflow: Workflow;

  constructor(
    public readonly project: TypeScriptProject,
    options: DeployJobOptions
  ) {
    const branchPrefix = options.branchPrefix ?? "main";
    const name = `Deploy ${options.appName} to ${options.environment.name}`;
    const jobId = options.jobId ?? paramCase(name);

    super(project, {
      jobId,
      name,
      ...options,
    });

    // defaults
    this.workflow = Workflow.deploy(project);
    this.branchPrefix = branchPrefix;
    this.needs = [this.workflow.buildJob.jobId, ...(options.needs ?? [])];
    this.deploySteps = options.deploySteps;
    this.environment = options.environment;
    this.artifactsDirectory = options.artifactsDirectory;

    // add this branch pattern as a trigger for the workflow
    this.workflow.addPushTrigger(this.branchPrefix);

    // make sure build step will store the artifacts we need
    this.workflow.buildJob.artifactDirectories.push(this.artifactsDirectory);
  }

  dependsOn = (job: WorkflowJob): void => {
    this.needs.push(job.jobId);
  };

  render = (): Job => {
    return {
      name: this.name,
      runsOn: this.workflow.defaultRunners,
      needs: this.needs,
      concurrency: this.jobId,
      if: `startsWith( github.ref, 'refs/heads/${this.branchPrefix}' )`,
      env: {
        CI: "true",
      },
      permissions: {
        contents: JobPermission.READ,
        ...(this.environment.authProviderType === "AWS_GITHUB_OIDC"
          ? {
              idToken: JobPermission.WRITE,
            }
          : {}),
      },
      steps: [
        ...renderDownloadArtifactSteps(this),
        ...renderSetupNodeSteps(this.workflow, {
          packageCache: false,
          setupPnpm: false,
        }),
        ...renderDeploySteps(this),
      ],
    };
  };
}
