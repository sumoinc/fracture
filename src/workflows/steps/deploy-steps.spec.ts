import { JobStep } from "projen/lib/github/workflows-model";
import { TypeScriptProject } from "projen/lib/typescript";
import { renderDeploySteps } from "./deploy-steps";
import { Environment } from "../../environments";
import { DeployJob } from "../jobs/deploy-job";

describe("success conditions", () => {
  test("Smoke test", () => {
    const project = new TypeScriptProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const deploySteps: Array<JobStep> = [
      {
        name: "Say foo",
        run: "echo 'foo'",
      },
    ];
    const environment = new Environment(project, {
      name: "my-environment",
    });
    const deployJob = new DeployJob(project, {
      deploySteps,
      environment,
      artifactsDirectory: "foo",
    });
    const outputSteps = renderDeploySteps(deployJob);
    expect(outputSteps).toBeTruthy();
    expect(outputSteps).toMatchSnapshot();
    // console.log(outputSteps);
  });

  test("With OIDC Configuration test", () => {
    const project = new TypeScriptProject({
      name: "my-project",
      defaultReleaseBranch: "main",
    });
    const deploySteps: Array<JobStep> = [
      {
        name: "Say foo",
        run: "echo 'foo'",
      },
    ];
    const environment = new Environment(project, {
      name: "my-environment",
    });
    const deployJob = new DeployJob(project, {
      deploySteps,
      environment,
      artifactsDirectory: "foo",
    });
    const outputSteps = renderDeploySteps(deployJob);
    expect(outputSteps).toBeTruthy();
    expect(outputSteps).toMatchSnapshot();
    // console.log(outputSteps);
  });
});
