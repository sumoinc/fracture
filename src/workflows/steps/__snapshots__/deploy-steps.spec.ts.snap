// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`success conditions Smoke test 1`] = `
Array [
  Object {
    "name": "Say foo",
    "run": "echo 'foo'",
  },
]
`;

exports[`success conditions With OIDC Configuration test 1`] = `
Array [
  Object {
    "name": "Say foo",
    "run": "echo 'foo'",
  },
]
`;
